bin/oneoff/backfill_data/backfill_script_ids_channel_tokens.rb

```diff
@@ -0,0 +1,211 @@
+# 4. If an associated user exists but the user has no script progress or user levels possibly associated
+# with the channel token, it's likely that the user visited the level page, which generated a channel token,
+# and then left the page. In this case, assign any of the possible scripts to the channel token
+
+require_relative '../../../dashboard/config/environment'
+require 'cdo/db'
+require 'optparse'
+require 'csv'
+
+# We started writing script_id to the table here: https://github.com/code-dot-org/code-dot-org/pull/39855
+# so we don't need to backfill past that point
+MAX_CHANNEL_TOKEN_ID_FOR_BACKFILL = 303_500_000
+
+# Parse options
+options = {
+  start_id: 1,
+  end_id: MAX_CHANNEL_TOKEN_ID_FOR_BACKFILL,
+  dry_run: false,
+}
+
+OptionParser.new do |opts|
+  opts.banner = <<~BANNER
+    Usage: #{File.basename(__FILE__)} [options]
+
+    This script backfills script_id on the Channel Tokens table.
+
+    Options:
+  BANNER
+
+  opts.on('--start-id=1234567',
+    Integer,
+    'Id of first entry to backfill (inclusive).'
+  ) do |start_id|
+    options[:start_id] = start_id
+  end
+
+  opts.on('--end-id=1234567',
+    Integer,
+    'Id of last entry to backfill (inclusive).'
+  ) do |end_id|
+    options[:end_id] = end_id
+  end
+
+  opts.on('--dry-run',
+    'Enables read-only mode where no changes are written to the database'
+  ) do |dry_run|
+    options[:dry_run] = dry_run
+  end
+
+  opts.on('-h', '--help', 'Prints this help message') do
+    puts opts
+    exit
+  end
+end.parse!
+puts "Options: #{options}"
+options.freeze
+
+# we're storing channel token ids that failed backfill in a csv for easy lookup later
+$csv_filename = "bin/oneoff/backfill_data/channel-token-backfill-failures.csv"
+$start_id = options[:start_id]
+$end_id = options[:end_id]
+$is_dry_run = options[:dry_run]
+
+$unable_to_backfill = 0
+
+$level_to_script_ids = {}
+
+def update_script_ids
+  puts "Backfilling channel token script_ids..."
+  puts "Script started at #{Time.now}"
+
+  CSV.open($csv_filename, "a") do |csv|
+    csv.sync = true
+
+    # find_each uses find_in_batches with a batch size of 1000 (https://apidock.com/rails/ActiveRecord/Batches/find_each)
+    ChannelToken.where(id: $start_id..$end_id).find_each do |channel_token|
+      next if channel_token.script_id.present?
+
+      begin
+        level = channel_token.level
+        user_id = user_id_for_storage_id(channel_token.storage_id)
+
+        # get all the possible scripts the channel_token could be associated with
+        associated_script_ids = get_associated_script_ids(level)
+
+        # if the level is associated with only one script_level, use that script
+        if associated_script_ids.length == 1
+          script_id = associated_script_ids[0]
+          update_channel_token(channel_token, script_id, csv)
+          next
+        end
+
+        # it's possible the channel token was generated for a logged-out user, in which case no user_id will exist
+        if user_id.present?
```

Comment:
(very minor) maybe just move to the next iteration if `user_id.nil?`.

Reply:
We'll still want to call `log_backfill_failed` for this case so I'm going to leave as-is